File: 3-HARD/Python/6-optimal-binary-search-tree/proposed-solution/optimal_bst.py
from typing import List

class OptimalBST:
    def __init__(self, keys: List[int], freq: List[int]):
        self.keys = keys
        self.freq = freq
        self.n = len(keys)
        self.dp = [[0] * self.n for _ in range(self.n)]

    def optimal_cost(self) -> int:
        # Precompute prefix sums for frequencies
        prefix_sum = [0] * (self.n + 1)
        for i in range(self.n):
            prefix_sum[i + 1] = prefix_sum[i] + self.freq[i]

        def get_sum(i, j):
            return prefix_sum[j + 1] - prefix_sum[i]

        def cost_for_root(i, j, r):
            left = self.dp[i][r - 1] if r > i else 0
            right = self.dp[r + 1][j] if r < j else 0
            return left + right + get_sum(i, j)

        for i in range(self.n):
            self.dp[i][i] = self.freq[i]

        for length in range(2, self.n + 1):
            for i in range(self.n - length + 1):
                j = i + length - 1
                self.dp[i][j] = min(
                    cost_for_root(i, j, r) for r in range(i, j + 1)
                )

        return self.dp[0][self.n - 1]

File: 3-HARD/JavaScript/6-optimal-binary-search-tree/proposed-solution/OptimalBST.js
// OptimalBST.js - Optimal Binary Search Tree using Dynamic Programming in JavaScript

class OptimalBST {
  constructor(keys, freq) {
    this.keys = keys;
    this.freq = freq;
    this.n = keys.length;
    this.dp = Array.from({ length: this.n }, () => new Array(this.n).fill(0));
  }

  optimalCost() {
    const prefixSum = new Array(this.n + 1).fill(0);
    for (let i = 0; i < this.n; i++) {
      prefixSum[i + 1] = prefixSum[i] + this.freq[i];
    }

    const getSum = (i, j) => prefixSum[j + 1] - prefixSum[i];

    const costForRoot = (i, j, r) => {
      const left = r > i ? this.dp[i][r - 1] : 0;
      const right = r < j ? this.dp[r + 1][j] : 0;
      return left + right + getSum(i, j);
    };

    for (let i = 0; i < this.n; i++) {
      this.dp[i][i] = this.freq[i];
    }

    for (let length = 2; length <= this.n; length++) {
      for (let i = 0; i <= this.n - length; i++) {
        const j = i + length - 1;
        let minCost = Infinity;
        for (let r = i; r <= j; r++) {
          const cost = costForRoot(i, j, r);
          if (cost < minCost) {
            minCost = cost;
          }
        }
        this.dp[i][j] = minCost;
      }
    }

    return this.dp[0][this.n - 1];
  }
}

export default OptimalBST;
